### PR TITLE
[WIP] Ad notifications fail to open if Brave is closed after the Ad notification is shown

### PR DESCRIPTION
**Pull request by tmancey**
## Submitter Checklist:

- [ ] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [ ] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- [ ] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: vendor/brave-ios/Ads/Generated/NativeAdsClient.h [hunk 1]
```diff
@@ -43,5 +42,6 @@ class NativeAdsClient : public ads::AdsClient {
   void SetIdleThreshold(const int threshold) override;
   uint32_t SetTimer(const uint64_t time_offset) override;
   void ShowNotification(std::unique_ptr<ads::NotificationInfo> info) override;
+  void CloseNotification(const std::string& id) override;
```
Comment (kylehickinson) [inline vendor/brave-ios/Ads/Generated/NativeAdsClient.h]: Needs an accompanying implementation in `vendor/brave-ios/Ads/Generated/NativeAdsClient.mm` which calls the Obj-C bridge method you added in `vendor/brave-ios/Ads/Generated/NativeAdsClientBridge.h`

File: vendor/brave-ios/Ads/BATBraveAds.mm [hunk 2]
```diff
@@ -454,11 +452,4 @@ - (void)eventLog:(const std::string &)json
   return std::make_unique<RewardsLogStream>(file, line, log_level);
 }
 
-#pragma mark - Misc
-
-- (const std::string)generateUUID
-{
-  return [self.commonOps generateUUID];
-}
-
```
Comment (kylehickinson) [inline vendor/brave-ios/Ads/BATBraveAds.mm]: Missing method added in `vendor/brave-ios/Ads/Generated/NativeAdsClientBridge.h`:

`- (void)closeNotification:(const std::string &)id;` 

You can implement this as an empty method here but please just leave a reference to what its for/this PR

File: vendor/bat-native-ads/include/bat/ads/notification_info.h [hunk 3]
```diff
@@ -16,14 +16,15 @@ namespace ads {
...
-  explicit NotificationInfo(const NotificationInfo& info);
+  NotificationInfo(const NotificationInfo& info);
   ~NotificationInfo();
 
   const std::string ToJson() const;
   Result FromJson(
       const std::string& json,
       std::string* error_description = nullptr);
 
+  std::string id;
```
Comment (kylehickinson) [inline vendor/bat-native-ads/include/bat/ads/notification_info.h]: Added a new field here but iOS has `BATAdsNotification` which needs the same field:

- [New Readonly Property](https://github.com/brave/brave-core/blob/a0e854a10668d2c55c7644f6d4149c79744102dc/vendor/brave-ios/Ads/BATAdsNotification.h#L19)
- [New Writable Property](https://github.com/brave/brave-core/blob/a0e854a10668d2c55c7644f6d4149c79744102dc/vendor/brave-ios/Ads/BATAdsNotification.mm#L10)
- [Updated initializer](https://github.com/brave/brave-core/blob/a0e854a10668d2c55c7644f6d4149c79744102dc/vendor/brave-ios/Ads/BATAdsNotification.mm#L23)

**Comment by tmancey** [comment]
Closing as superseded by https://github.com/brave/brave-core/pull/2876

**Pull request closed without merging**